packages/react/src/useTable.tsx

```diff
@@ -248,10 +149,105 @@ const getRows = (props: { data: GadgetRecord<any>[]; columns: TableColumn[] }) =
   const { columns, data } = props;
 
   return data.map((record) => {
-    const row: Record<string, ColumnValueType> = { id: record.id };
+    const row: TableRow = { id: record.id };
     for (const { apiIdentifier, getValue } of columns) {
       row[apiIdentifier] = getValue(record);
     }
     return row;
   });
 };
+
+const getTableColumnArray = (fields: FieldMetadataFragment[], columnsFromOptions: TableOptions["columns"]) => {
+  const columnsMap =
+    columnsFromOptions &&
+    new Map(
+      columnsFromOptions.map((column) => {
+        if (isRelatedFieldColumn(column)) {
+          return [column.field, column.relatedField];
+        } else {
+          return [column, undefined];
+        }
+      })
+    );
+
+  const result: TableColumn[] = fields.map((field) => {
+    let relatedField: RecordTableColumnValue | undefined;
+    const relatedFieldColumn = columnsMap?.get(field.apiIdentifier);
+
+    if (relatedFieldColumn) {
```

Comment:
This entire piece of logic for getting the related field feels like it could be pulled into a helper function